File: 백준/Gold/15683. 감시/감시.py
import sys

def dir(arr, k):
    if k == 1:
        return arr
    elif k == 2:
        return [arr[0] | arr[2], arr[1] | arr[3]]
    elif k == 3:
        return [arr[0] | arr[1], arr[1] | arr[2],  arr[2] | arr[3] , arr[3] | arr[0] ]
    elif k == 4:
        return [arr[0] | arr[1] | arr[2], arr[1] | arr[2] | arr[3], arr[2] | arr[3] | arr[0], arr[3] | arr[0] | arr[1]]
    elif k == 5:
        return [arr[1] | arr[3] | arr[0] | arr[2]]

def make_set(i,j, pans,n,m):
    arr = []
    for a,b in [[0,1],[1,0],[0,-1],[-1,0]]:
        k = 0
        s = set()
        while True:
            if i + a*k >= n or j + b*k >= m or i + a*k < 0 or j + b*k < 0 or pans[i+ a*k][j + b*k] == 6:
                break;
            if pans[i+ a*k][j + b*k] == 0:
                s.add((i+ a*k, j + b*k))
            k += 1
        arr.append(s)
    return dir(arr, pans[i][j]);

def solution(n, m, pans):

    result = 0;
    k = 0;
    block = 0
    que = []
    for i in range(n):
        for j in range(m):
            if pans[i][j] in [1,2,3,4,5]:
                k += 1;
                que.append((i,j))
            elif pans[i][j] == 6:
                block += 1
    boards = [[] for _ in range(k)]
    if len(que) == 0:
        return n*m - block

    for i in range(k):
        s_list = make_set(que[i][0], que[i][1], pans,n,m)
        for s in s_list:
            for b in (boards[i-1] if i > 0 else [set()]):
                new_s = s | b
                boards[i].append(new_s)

    for i in range(len(boards[k-1])):
        result = max(result, len(boards[k-1][i]))
    return n*m - len(que) - block - result

n,m = map(int,sys.stdin.readline().split());
myList = [list(map(int,sys.stdin.readline().split(' '))) for _ in range(n)];
print(solution(n, m, myList));

File: 백준/Gold/15683. 감시/감시.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Main {

  private static BufferedReader br;
  private static StringBuilder sb = new StringBuilder();


  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));

    String[] line = br.readLine().split(" ");
    int n = Integer.parseInt(line[0]);
    int m = Integer.parseInt(line[1]);
    int[][] pan = new int[n][m];
    for (int i = 0; i < n; i++) {
      String[] split = br.readLine().split(" ");
      for(int j = 0 ; j < m ; j++){
        pan[i][j] = Integer.parseInt(split[j]);
      }
    }

    System.out.println(solution(n, m, pan));
  }


  public static int solution(int n ,int m, int[][] arr) {
    int result = Integer.MAX_VALUE;
    List<Node> cctvs = new ArrayList<>();

    for(int i = 0; i < n ; i++){
      for(int j = 0 ; j < m ; j++){
        if(arr[i][j] > 0 && arr[i][j] < 6){
          cctvs.add(new Node(i, j, arr[i][j]));
        }
      }
    }

    int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    Queue<QN> queue = new LinkedList<>();
    queue.add(new QN(0, arr));

    while(!queue.isEmpty()){
      QN qn = queue.poll();
      int idx = qn.idx;
      int[][] prevMap = qn.map;


      if(idx == cctvs.size()){
        int cnt = 0;
        for(int i = 0 ; i < n ; i++){
          for(int j = 0 ; j < m ; j++){
            if(prevMap[i][j] == 0){
              cnt++;
            }
          }
        }
        result = Math.min(result, cnt);
        continue;
      }
      int type = cctvs.get(idx).type;
      if(type == 1){
        for(int i = 0 ; i < 4 ; i++){
          int[][] newMap = deepCopy(n, m, prevMap);
          int[] dir = dirs[i];
          for(int j = 1 ; j < 8 ; j++){
            int nx = cctvs.get(idx).x + dir[0] * j;
            int ny = cctvs.get(idx).y + dir[1] * j;
            if(nx < 0 || ny < 0 || nx >= n || ny >= m || newMap[nx][ny] == 6){
              break;
            } else if (newMap[nx][ny] > 0 && newMap[nx][ny] < 6){
              continue;
            }
            newMap[nx][ny] = 7;
          }
          queue.add(new QN(idx + 1, newMap));
        }
      } else if (type == 2){
        for(int k = 0; k < 2 ; k++) {
          int[][] newMap = deepCopy(n, m, prevMap);
          for (int i = 0; i < 2; i++) {
            int[] dir = dirs[i*2 + k];
            for (int j = 1; j < 8; j++) {
              int nx = cctvs.get(idx).x + dir[0] * j;
              int ny = cctvs.get(idx).y + dir[1] * j;
              if (nx < 0 || ny < 0 || nx >= n || ny >= m || newMap[nx][ny] == 6) {
                break;
              } else if (newMap[nx][ny] > 0 && newMap[nx][ny] < 6) {
                continue;
              }
              newMap[nx][ny] = 7;
            }
          }
          queue.add(new QN(idx + 1, newMap));
        }
      } else if (type == 3){
        for(int k = 0; k < 4 ; k++) {
          int[][] newMap = deepCopy(n, m, prevMap);
          for (int i = k; i < k+2; i++) {
            int[] dir = dirs[(i)%4];
            for (int j = 1; j < 8; j++) {
              int nx = cctvs.get(idx).x + dir[0] * j;
              int ny = cctvs.get(idx).y + dir[1] * j;
              if (nx < 0 || ny < 0 || nx >= n || ny >= m || newMap[nx][ny] == 6) {
                break;
              } else if (newMap[nx][ny] > 0 && newMap[nx][ny] < 6) {
                continue;
              }
              newMap[nx][ny] = 7;
            }
          }
          queue.add(new QN(idx + 1, newMap));
        }
      } else if (type == 4){
        for(int k = 0; k < 4 ; k++) {
          int[][] newMap = deepCopy(n, m, prevMap);
          for (int i = k; i < k+3; i++) {
            int[] dir = dirs[(i)%4];
            for (int j = 1; j < 8; j++) {
              int nx = cctvs.get(idx).x + dir[0] * j;
              int ny = cctvs.get(idx).y + dir[1] * j;
              if (nx < 0 || ny < 0 || nx >= n || ny >= m || newMap[nx][ny] == 6) {
                break;
              } else if (newMap[nx][ny] > 0 && newMap[nx][ny] < 6) {
                continue;
              }
              newMap[nx][ny] = 7;
            }
          }
          queue.add(new QN(idx + 1, newMap));
        }
      } else if (type == 5){
        int[][] newMap = deepCopy(n, m, prevMap);
        for(int i = 0 ; i < 4 ; i++){
          int[] dir = dirs[i];
          for(int j = 1 ; j < 8 ; j++){
            int nx = cctvs.get(idx).x + dir[0] * j;
            int ny = cctvs.get(idx).y + dir[1] * j;
            if(nx < 0 || ny < 0 || nx >= n || ny >= m || newMap[nx][ny] == 6){
              break;
            } else if (newMap[nx][ny] > 0 && newMap[nx][ny] < 6){
              continue;
            }
            newMap[nx][ny] = 7;
          }
        }
        queue.add(new QN(idx + 1, newMap));
      }

    }

    return result;
  }

  private static int[][] deepCopy(int n, int m, int[][] prevMap) {
    int[][] newMap = new int[n][m];
    for(int j = 0 ; j < n; j++){
      newMap[j] = Arrays.copyOf(prevMap[j], m);
    }
    return newMap;
  }


  public static class QN{
    int idx;
    int[][] map;
    public QN(int idx, int[][] map){
      this.idx = idx;
      this.map = map;
    }
  }


  public static class Node{
    int x;
    int y;
    int type;

    public Node(int x, int y, int type){
      this.x = x;
      this.y = y;
      this.type = type;
    }

    @Override
    public boolean equals(Object o){
      if(o instanceof Node){
        Node node = (Node) o;
        return this.x == node.x && this.y == node.y;
      }
      return false;
    }

    @Override
    public int hashCode(){
      return x * 1000 + y;
    }
  }
}
